packages/runed/src/lib/utilities/useNetworkStatus/useNetworkStatus.svelte.ts

```diff
@@ -0,0 +1,145 @@
+
+interface NetworkStatus {
+	/**
+	 * @desc Returns the online status of the browser.
+	 * @see https://developer.mozilla.org/en-US/docs/Web/API/Navigator/onLine
+	 */
+	online: boolean;
+	/**
+	 * @desc The {Date} object pointing to the moment when state update occurred.
+	 */
+	updatedAt: Date;
+	/**
+	 * @desc Effective bandwidth estimate in megabits per second, rounded to the
+	 * nearest multiple of 25 kilobits per seconds.
+	 * @see https://developer.mozilla.org/en-US/docs/Web/API/NetworkInformation/downlink
+	 */
+	downlink?: NetworkInformation["downlink"];
+	/**
+	 * @desc Maximum downlink speed, in megabits per second (Mbps), for the
+	 * underlying connection technology
+	 * @see https://developer.mozilla.org/en-US/docs/Web/API/NetworkInformation/downlinkMax
+	 */
+	downlinkMax?: NetworkInformation["downlinkMax"];
+	/**
+	 * @desc Effective type of the connection meaning one of 'slow-2g', '2g', '3g', or '4g'.
+	 * This value is determined using a combination of recently observed round-trip time
+	 * and downlink values.
+	 * @see https://developer.mozilla.org/en-US/docs/Web/API/NetworkInformation/effectiveType
+	 */
+	effectiveType?: NetworkInformation["effectiveType"];
+	/**
+	 * @desc Estimated effective round-trip time of the current connection, rounded
+	 * to the nearest multiple of 25 milliseconds
+	 * @see https://developer.mozilla.org/en-US/docs/Web/API/NetworkInformation/rtt
+	 */
+	rtt?: NetworkInformation["rtt"];
+	/**
+	 * @desc {true} if the user has set a reduced data usage option on the user agent.
+	 * @see https://developer.mozilla.org/en-US/docs/Web/API/NetworkInformation/saveData
+	 */
+	saveData?: NetworkInformation["saveData"];
+	/**
+	 * @desc The type of connection a device is using to communicate with the network.
+	 * It will be one of the following values:
+	 *  - bluetooth
+	 *  - cellular
+	 *  - ethernet
+	 *  - none
+	 *  - wifi
+	 *  - wimax
+	 *  - other
+	 *  - unknown
+	 *  @see https://developer.mozilla.org/en-US/docs/Web/API/NetworkInformation/type
+	 */
+	type?: NetworkInformation["type"];
+}
+
+/**
+ * Tracks the state of browser's network connection.
+ * @returns null if the function is not run in the browser
+ */
+export function useNetworkStatus() {
+	if (!browser) {
+		return {
+			get current() {
+				return null;
+			},
+			get previous() {
+				return null;
+			},
+		};
+	}
+	const navigator = window.navigator;
+	const connection =
+		"connection" in navigator ? (navigator as NavigatorWithConnection).connection : undefined;
+	const getConnectionStatus = (): NetworkStatus => {
+		return {
+			online: navigator.onLine,
+			updatedAt: new Date(),
+			downlink: connection?.downlink,
+			downlinkMax: connection?.downlinkMax,
+			effectiveType: connection?.effectiveType,
+			rtt: connection?.rtt,
+			saveData: connection?.saveData,
+			type: connection?.type,
+		};
+	};
+
+	let current = $state(getConnectionStatus());
+	const previous = new Previous(() => current);
```

Comment:
nitpick/discussion: I question whether previous is worth the overhead to always be automatically included here, or just something you explicitly choose to track with `Previous()` when needed.

i.e.
```ts
const networkStatus = useNetworkStatus();
const previousNetworkStatus = new Previous(() => networkStatus.current);
```

Reply:
I find it useful to have it included, because I could imagine some real use-cases that rely on that. 

My perspective is that as a developer choosing a util library for a project, I may not be familiar with all of its independent features right away, especially if I'm only focusing on a few at a time. If a standalone function can solve most of my use cases without needing other imports, I can overlook the overhead.

I know it's very subjective, though, and I definitely see your point too. My opinion is not very strong in this case, so I can let it go, but then I'd like to make sure this example is included in the docs. :)